### PR TITLE
[WIP] Services IPFamily integration tests

### PR DESCRIPTION
**Pull request by aojea**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

As stated here https://github.com/kubernetes/kubernetes/pull/86895, 
alpha features are prone to have errors if they are not well tested 
running skewed API servers.

To graduate the dual-stack feature to beta we should add skew testing

ref: https://github.com/kubernetes/kubernetes/issues/81035

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: test/integration/network/dual_stack_test.go [hunk 1]
```diff
@@ -0,0 +1,336 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (aojea) [inline test/integration/network/dual_stack_test.go]: 2020

File: test/integration/network/dual_stack_test.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+	clientSet, _ := framework.StartTestServer(t, stopCh, framework.TestServerSetup{
+		ModifyServerRunOptions: func(opts *options.ServerRunOptions) {
+			opts.ServiceClusterIPRanges = "10.0.0.0/16"
+		},
+		ModifyServerConfig: func(config *master.Config) {
+		},
+	})
+
+	ipFamily := v1.IPv4Protocol
+	svc := createService(&ipFamily, v1.ServiceTypeClusterIP)
```
Comment (aojea) [inline test/integration/network/dual_stack_test.go]: this block is being duplicated, I think that I will create a table driven test with the following parameters:
- ServiceClusterIPRanges
- ipFamily

File: test/integration/network/dual_stack_test.go [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+	for _, s := range svcs.Items {
+		if err = validateServiceAndClusterIPFamily(&s, ipFamily); err != nil {
+			t.Fatalf("unexpected service IP family %s: %v", s.Spec.ClusterIP, err)
+		}
+	}
+
+}
+
+// TestDualStackSkew tests two dual stack API server with and without DualStack enabled
+func TestDualStackSkew(t *testing.T) {
```
Comment (aojea) [inline test/integration/network/dual_stack_test.go]: @liggitt  I'll need help here to understand how to "create" this testing
https://github.com/kubernetes/kubernetes/issues/81035

File: test/integration/network/dual_stack_test.go [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+	netutils "k8s.io/utils/net"
+)
+
+// TestDualStackEnabledSingleIPv4Family tests a single IPv4 family API server with DualStack enabled
+func TestDualStackEnabledSingleIPv4Family(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.IPv6DualStack, true)()
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	clientSet, _ := framework.StartTestServer(t, stopCh, framework.TestServerSetup{
```
Comment (aojea) [inline test/integration/network/dual_stack_test.go]: after checking several integrations tests, seems there is no consistency on how to start a kube-apiserver, or I'm missing the goal of using one or another way, 
what's the recommended way of starting a kube-apiserver for these tests?

Comment (liggitt) [reply test/integration/network/dual_stack_test.go]: for something like this where you want to exercise a variety of invocation options, and don't care about controller-manager running, I'd probably use

```go
import (
  kubeapiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
  "k8s.io/kubernetes/test/integration/framework"
)

...

kubeapiservertesting.StartTestServerOrDie(t, nil, []string{
  ... apiserver flags go here ...
}, framework.SharedEtcd())
```

Comment (liggitt) [reply test/integration/network/dual_stack_test.go]: that lets you explicitly enable/disable the feature with a `--feature-gates` arg, and explicitly set up an IPv4 or IPV6 or dual-stack API server configuration

File: test/integration/network/services_test.go [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+			if tc.dualStack && err != nil {
+				t.Errorf("unexpected error text: %v", err)
+			}
+		})
+	}
+}
+
+// TestTransitionService ClusterIP -> Headless -> External -> clear ClusterIP -> ClusterIP
+// TestServicesIPFamily tests the IP family field on DualStack clusters
+func TestTransitionService(t *testing.T) {
```
Comment (aojea) [inline test/integration/network/services_test.go]: need to go deeper into this test but this is the output I'm getting
```
    --- FAIL: TestTransitionService/DualStack_enabled_and_IPv6_IPv4 (4.80s)
        testserver.go:181: runtime-config=map[api/all:true]
        testserver.go:182: Starting kube-apiserver on port 52840...
        testserver.go:198: Waiting for /healthz to be ok...
        services_test.go:294: unexpected error text: Service "svc-transition" is invalid: [spec.externalName: Required value, spec.ipFamily: Required value]
        services_test.go:294: unexpected error text: Service "svc-transition" is invalid: [spec.clusterIP: Invalid value: "": field is immutable, spec.ipFamily: Invalid value: "null": field is immutable, spec.ipFamily: Required value]
        services_test.go:294: unexpected error text: Service "svc-transition" is invalid: [spec.clusterIP: Invalid value: "": field is immutable, spec.ipFamily: Invalid value: "null": field is immutable, spec.ipFamily: Required value]
        services_test.go:294: unexpected error text: Service "svc-transition" is invalid: [spec.clusterIP: Invalid value: "": field is immutable, spec.ipFamily: Invalid value: "null": field is immutable, spec.ipFamily: Required value]
```

File: test/integration/network/services_test.go [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+			if _, err := client.CoreV1().Services(metav1.NamespaceDefault).Create(svc); err != nil {
+				t.Errorf("unexpected error text: %v", err)
+			}
+
+			// Create a service without IPFamily and check that it's created with the default IP family
+			svc = createService("svc-ipfamily-nil", nil, v1.ServiceTypeClusterIP)
+			if _, err := client.CoreV1().Services(metav1.NamespaceDefault).Create(svc); err != nil {
+				t.Errorf("unexpected error text: %v", err)
+			}
+
```
Comment (liggitt) [inline test/integration/network/services_test.go]: also exercise these cases?
* type=ExternalName, ipFamily=nil
* type=ExternalName, ipFamily=default
* type=ExternalName, ipFamily=non-default (should succeed in dual-stack cluster, fail in single-stack cluster)
* type=ClusterIP, clusterIP=None, ipFamily=nil
* type=ClusterIP, clusterIP=None, ipFamily=default
* type=ClusterIP, clusterIP=None, ipFamily=non-default

File: test/integration/network/services_test.go [hunk 7]
```diff
@@ -0,0 +1,455 @@
...
+			// Transition the service over all the types of services and check that the IPFamily is inmutable
+			ipFamily := tc.defaultIPFamily
+
+			svcType := []v1.ServiceType{
+				v1.ServiceTypeExternalName,
+				v1.ServiceTypeNodePort,
+				v1.ServiceTypeLoadBalancer,
+				v1.ServiceTypeClusterIP,
+				v1.ServiceTypeClusterIP, // no-op
+			}
```
Comment (liggitt) [inline test/integration/network/services_test.go]: we need a second dimension that simulates a client that is unaware of the new IPFamily field and drops it (sets it to nil) on every update

File: test/integration/network/services_test.go [hunk 8]
```diff
@@ -0,0 +1,455 @@
...
+					}
+				} else if err = validateServiceAndClusterIPFamily(s, ipFamily); err != nil {
+					t.Fatalf("unexpected service IP family %s: %v", s.Spec.ClusterIP, err)
+				}
+
+				resourceVersion, err := strconv.ParseFloat(s.ResourceVersion, 64)
+				if err != nil {
+					t.Errorf("error converting resource version to int: %v", err)
+				}
+				resourceVersion++
```
Comment (liggitt) [inline test/integration/network/services_test.go]: why are you incrementing the resourceVersion? the update should be sent with the same version you received from the Get

File: test/integration/network/services_test.go (hunk 5 above)
Comment (liggitt) [reply test/integration/network/services_test.go]: the first error is because you are trying to set type=externalname type without specifying the required externalName field.

the other errors might be related to the bug @smarterclayton is trying to fix, I'm not sure

File: test/integration/network/services_test.go (hunk 8 above)
Comment (aojea) [reply test/integration/network/services_test.go]: I did that and it throwed this error ```metadata.resourceVersion: Invalid value: "": must be specified for an update``` and this was the answer I've found googling around.
Maybe the error was related to the other fields? let me try again

File: test/integration/network/main_test.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
+/*
+<<<<<<< HEAD
+<<<<<<< HEAD
```
Comment (aojea) [inline test/integration/network/main_test.go]: 🤦‍♂

**Comment by neolit123** [comment]
/remove-area apiserver cloudprovider code-generation conformance dependency e2e-test-framework kubeadm kubectl
/remove-sig api-machinery apps architecture auth cli cloud-provider cluster-lifecycle instrumentation node

**Comment by aojea** [comment]
@smarterclayton @liggitt 
I've simplified the testing to check only the service creation and the IPfamily behavior on a cluster with two skewed apiservers, currently, this is failing because of:

- It can create services with any IPFamily on single stack clusters, should it error?
- A client connected to an apiserver without dual-stack enabled returns the IPFamily field for all the services, my understanding is that it should return nil

It may be possible that I'm doing something wrong?

File: test/integration/network/services_ipfamily_test.go [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+
+				// cleanup the registry storage
+				defer registry.CleanupStorage()
+
+				// start api server without dual stack enabled
+				server := kubeapiservertesting.StartTestServerOrDie(t, instanceOptions, []string{
+					"--endpoint-reconciler-type", "master-count",
+					"--advertise-address", "10.0.1.1",
+					"--apiserver-count", "1",
+					"--feature-gates", "IPv6DualStack=false",
```
Comment (liggitt) [inline test/integration/network/services_ipfamily_test.go]: the servers use a global feature registry... you can't run two servers with different settings for the feature gate in the same process

Comment (aojea) [reply test/integration/network/services_ipfamily_test.go]: aaargh, :face_with_head_bandage: 

so ... it doesn't seem feasible from the integration framework to do this test ...
:thinking:  how can I simulate a client that's unaware of the ipFamily?

**Comment by aojea** [comment]
outdated and superseded by the new KEP
